### ISSUE TITLE
FindTinyXML.cmake is missing _LIBRARY_DIRS

### ISSUE THREAD

**Issue by isucan** [ros/cmake_modules#5]
If tinyxml is not installed in a default location, the linking will fail without calling link_directories(). The .cmake module should set the _LIBRARY_DIRS variable as well.

**Comment by wjwwood** [comment]
I'm not sure that's true, I would expect `TinyXML_LIBRARIES` to alway contain absolute paths to the libraries.

Reading over the module, it seems like it might not find TinyXML if it is in a non-standard place unless you set the `TinyXML_ROOT_DIR`, but the use of `_LIBRARY_DIRS` shouldn't be required.

@adolfo-rt care to comment since you contributed the module originally?

**Comment by adolfo-rt** [comment]
> > If tinyxml is not installed in a default location, the linking will fail without calling link_directories(). The .cmake module should set the _LIBRARY_DIRS variable as well.

If the TinyXML installation lives in a nonstandard location and this location is not specified anywhere, the find script will fail and all output variables should be empty (except `TinyXML_FOUND == false`). Exposing `TinyXML_LIBRARY_DIRS` would not help here. 

Some questions that might help diagnose what's going on:

What is `TinyXML_FOUND` evaluating to? If it evaluates to **true**, then `TinyXML_LIBRARIES` and all other output variables should contain valid information, otherwise consider this a bug.

How are you calling find_package?. If the dependency is required, you should be doing

```
find_package(TinyXML REQUIRED)
```

if OTOH it's an optional dependency, you should do something like

```
find_package(TinyXML)

if (TinyXML_FOUND)
  # Enable optional functionality that leverages TinyXML
endif()
```

What should **never** be done is:

```
find_package(TinyXML) # NOTE: No REQUIRED

add_executable(foo foo.cpp)

# We don't know if TinyXML was found or not. If not, output variables will be empty
target_link_libraries(foo ${TinyXML_LIBRARIES}) # Bad!
```

> I'm not sure that's true, I would expect TinyXML_LIBRARIES to alway contain absolute paths to the libraries.

Yes, it should contain fully qualified paths.

> Reading over the module, it seems like it might not find TinyXML if it is in a non-standard place unless you set the >TinyXML_ROOT_DIR, but the use of _LIBRARY_DIRS shouldn't be required.

How is CMake being informed about the nonstandard location of TinyXML?. I see these options:
- [CMAKE_PREFIX_PATH](http://www.cmake.org/Wiki/CMake_Useful_Variables). This is a global CMake setting that will affect all `find_*` commands.
- `TinyXML_ROOT_DIR` CMake or environment variable. Has a package-local effect.
- Go full manual, not recommended. Let the cmake step fail, fire up ccmake or cmake-gui, manually point `TinyXML_INCLUDE_DIR` and `TinyXML_LIBRARY` cache variables to their correct locations, and reconfigure.

**Comment by isucan** [comment]
Sorry about this, I should have been a bit more explicit. Things actually work fine for me.
Since this is the first module in the repo, I expect others will follow the style for providing their own cmake modules.
The question is, which cmake variables do we expect to have always defined and which are optional?

I would put the mandatory list to be:
- _FOUND
- _INCLUDE_DIRS
- _LIBRARY_DIRS
- _LIBRARIES

and optional:
- _DEFINITIONS

_LIBRARIES could also be made optional I guess if an absolute path is set to _LIBRARIES. Do you by any chance know if this is supported by Visual Studio?

I guess this is more of a documentation issue. Perhaps this would be better addressed by adding a 'how to contribute' section to http://ros.org/wiki/cmake_modules or to the README in this repo.

**Comment by adolfo-rt** [comment]
Currently the guidelines on [contributing](https://github.com/ros/cmake_modules/blob/master/CONTRIBUTING.md) state that required output variables should be (quote):
- <PackageName>_FOUND
- <PackageName>_INCLUDE_DIRS
- <PackageName>_LIBRARIES
- It may provide other variables but they must be documented at the top of the module file.

This sounds good enough for most cases. Being really pedantic, I'd argue that `<PackageName>_FOUND` should be the only required variable, eg. consider the cases ofFindDoxygen.cmake or FindPkgConfig.cmake, where you're mostly interested in finding an executable `<PackageName>_EXECUTABLE`.

Concerning `<PackageNames>_LIBRARY_DIRS`, I advise against using it. It's there mostly for legacy reasons, and is prone causing name resolution ambiguities. Using `<PackageName>_LIBRARIES` with fully qualified names gives no room for ambiguity.

So, how about rewording the required variables section in CONTRIBUTING.md to:

The module should at least provide the `<PackageName>_FOUND` variable, and when applicable, any of the following additional variables:
- `<PackageName>_INCLUDE_DIRS`
- `<PackageName>_LIBRARIES`
- `<PackageName>_EXECUTABLE`
- `<PackageName>_DEFINITIONS`

The most common usecase for writing CMake module scripts is that of finding libraries, case in which the output variables are:
- `<PackageName>_FOUND`
- `<PackageName>_INCLUDE_DIRS`
- `<PackageName>_LIBRARIES`

**Comment by isucan** [comment]
that sounds good. sorry for missing the contribution directions.
